File: 백준/Silver/28215. 대피소/대피소.py
from itertools import combinations

def mii():
    return map(int, input().split())

N, K = mii()
INF = float("INF")
x = [0] * N
y = [0] * N
for i in range(N):
    x[i], y[i] = mii()

def max_dist(comb):
    rst = 0
    for i in range(N):
        min_dist = INF
        for c in comb:
            dist = abs(x[i] - x[c]) + abs(y[i] - y[c])
            min_dist = min(min_dist, dist)
        rst = max(rst, min_dist)
    return rst

ans = INF
for comb in combinations(range(N), K):
    ans = min(ans, max_dist(comb))

print(ans)

File: 백준/Silver/28215. 대피소/대피소.cc
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;
int n, k;
int x[50], y[50];
int calc_dist(int i, int j, int k){
    int res = 0;
    for(int h=0; h<n; h++){
        res = max(res, min(min(abs(x[i]-x[h])+abs(y[i]-y[h]), abs(x[j]-x[h])+abs(y[j]-y[h])), 
                  abs(x[k]-x[h])+abs(y[k]-y[h])));
    }
    return res;
}

int main(){
    int ans = 1e9;
    cin >> n >> k;
    for(int i=0; i<n; i++){
        cin >> x[i] >> y[i];
    }
    if(k==1){
        for(int i=0; i<n; i++){
            ans = min(ans, calc_dist(i, i, i));
        }
    }
    else if(k==2){
        for(int i=0; i<n; i++){
            for(int j=0; j<n; j++){
                ans = min(ans, calc_dist(i, j, j));
            }
        }
    }
    else{
        for(int i=0; i<n; i++){
            for(int j=0; j<n; j++){
                for(int k=0; k<n; k++){
                    ans = min(ans, calc_dist(i, j, k));
                }
            }
        }
    }
    cout << ans;
}